### DISCUSSION TITLE
How can I add external C++ libraries to use in MOOSE?

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#28769] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

First I want to know generally how I would add an external library and add it to the Makefile so I can compile. Whether autodiff (what I'm asking about) or any other library in general. Also curious if there is a way to write a python script which is called, does the calculations I want, and returns the results to the C++ source file.

Second, I want to know how to add autodiff (https://github.com/autodiff/autodiff) and wondering if it necessary since MOOSE already has some kind of AD system that is supposed to do the same thing? I'm not sure. In my custom Material file I am currently calculating derivatives manually with finite differencing, for example:

```
// Compute the numerical derivative of a scalar function w.r.t a tensor_ij

RankTwoTensor ComputeStrainEnergyNeoHookeanNearlyIncompressible::compute_dWdC(const Real &mu_0, const RankTwoTensor &C) 
{
	// Initialize perturbation
	Real epsilon = 1e-6;
	
	// Initialize derivative
	RankTwoTensor dWdC;
	
    // Compute the scalar function for the original tensor
    Real W = computeStrainEnergy(mu_0, C);

    // Loop over all components of the tensor to compute the derivative
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            // Create a copy of the tensor to perturb
            RankTwoTensor C_perturbed = C;

            // Perturb the (i, j) component by epsilon
            C_perturbed(i, j) += epsilon;

            // Evaluate the scalar function at the perturbed tensor
            Real W_perturbed = computeStrainEnergy(mu_0, C_perturbed);

            // Compute the finite difference derivative and store it in the dfdT tensor
            dWdC(i, j) = (W_perturbed - W) / epsilon;
        }
    }

    // Return the tensor of derivatives
    return dWdC;
}

```

And I am getting good results, but I am wondering if the convergence and accuracy is affected by this because it solves a bit slowly and my choice of epsilon is problem-dependent. I want to instead add autodiff (https://github.com/autodiff/autodiff) do something like (I used chatGPT for this sample but I want to do something similar):

```
RankTwoTensor ComputeStrainEnergyNeoHookeanNearlyIncompressible::compute_dWdC(const Real &mu_0, const RankTwoTensor &C)
{
    using namespace autodiff;
    using dual = dual2nd;

    Eigen::Matrix<dual, 3, 3> C_dual;
    for (int i = 0; i < 3; ++i)
        for (int j = 0; j < 3; ++j)
            C_dual(i, j) = C(i, j);

    auto W = [&](const Eigen::Matrix<dual, 3, 3> & C_dual) {
        dual J = std::pow(C_dual.determinant(), 1.0 / 2.0);
        Eigen::Matrix<dual, 3, 3> C_iso = std::pow(J, -2.0 / 3.0) * C_dual;
        dual I_1 = C_iso.trace();
        return (mu_0 / 2.0) * (I_1 - 3);
    };

    Eigen::Matrix<dual, 3, 3> dWdC_dual = gradient(W, wrt(C_dual), at(C_dual));
    RankTwoTensor dWdC;
    for (int i = 0; i < 3; ++i)
        for (int j = 0; j < 3; ++j)
            dWdC(i, j) = val(dWdC_dual(i, j));  // Extract the value from dual

    return dWdC;
}
```

How can I do this? I have already tried https://mooseframework.inl.gov/source/materials/DerivativeMaterialInterface.html and it didn't work to calculate derivatives of scalars wrt tensors

**Comment by GiudGiud** [comment]
Hello

> First I want to know generally how I would add an external library and add it to the Makefile so I can compile. 

If the library is already compiled on your machine, you'll want to add to several environment variables in the Makefile.
However, I dont think you need to do that here.

We have automatic differentiation in moose using MetaPhyscL through libmesh. Every single object in MOOSE prefaced with AD already uses this. And all you need to do to use it is:

- inherit from the ADMaterial class for your custom material
- everywhere you have Real, use ADReal, same for RealVectorValue, RealTensorValue etc

**Reply by richmondodufisan** [reply]
okay thanks. could you please clarify how to use these derivatives? for example, I say I updated my function to calculate W to be:

```
ADReal ComputeStrainEnergyNeoHookeanNearlyIncompressible::computeStrainEnergy(const ADReal &mu_0, const ADRankTwoTensor &C)
{
	ADReal J = std::pow(C.det(), 1.0 / 2.0);
	
	// Isochoric Right Cauchy-Green deformation tensor
	ADRankTwoTensor C_iso = std::pow(J, -2.0 / 3.0) * C;
	
	// First invariant, equivalent to lambda_1^2 + lambda_2^2 + lambda_3^2
	ADReal I_1 = C_iso.trace();
	
	ADReal strain_energy = (mu_0/2.0) * (I_1 - 3);
	
	return strain_energy;
}
```

Now I want to calculate dW/dC, and use it *within the same material* to calculate the PK2 stress (=dW/dC + some other term).

Through trial and error I found that there is some "`MetaPhyscL::gradient`" but I cannot find the documentation anywhere online. Would it be something like

```
ADReal W = computeStrainEnergy(mu_0, C);

ADRankTwoTensor dWdC = MetaPhyscL::gradient(W, C);
```

Or how would I do this?

**Reply by GiudGiud** [reply]
> Now I want to calculate dW/dC, and use it within the same material to calculate the PK2 stress (=dW/dC + some other term).

you usually do not need to anymore. in the residual of your equation, if you needed dW/dC to compute the Jacobian, it would be retrieved automatically from the W term in the equation. 

But if you do still need to compute it, then this can be done:
```
    const auto dof_number = _current_elem->dof_number(_sys.number(), _C.number(), 0);
    const auto d_W_d_C = W.derivatives()[dof_number];
```
- C must be the variable (a mooseVariableFE type, which can be obtained with getVar("name")
- W is the ADReal property

gradient is only for spatial derivatives

**Reply by richmondodufisan** [reply]
Okay so let me give a bit of background on what I'm trying to do because I'm still a little confused. C in this case is not a mooseVariableFE, it is not even declared as a material property. It's just a C++ RankTwoTensor within this Material file.

Within this Material file [ComputeStrainEnergyNeoHookeanNearlyIncompressible](https://github.com/richmondodufisan/purple/blob/main/src/kernels/ComputeStrainEnergyNeoHookeanNearlyIncompressible.C) I want to calculate and declare:

-the PK2 stress, S
-the Jacobian, dS/dE

Then I have another Material file, [ComputeStressNeoHookean](https://github.com/richmondodufisan/purple/blob/main/src/kernels/ComputeStressNeoHookean.C) that inherits from [ComputeLagrangianStressPK2](https://mooseframework.inl.gov/source/materials/lagrangian/ComputeLagrangianStressPK2.html). It is in this file that I "get" the previous declared quantities and pass them for use in the stress divergence kernels. As I understand it, we are REQUIRED to provide MOOSE with both of these in the new solid mechanics module.

My since I am solving a nearly incompressible problem, I need to define my stress my way/a bit differently. 

In the first Material file,  [ComputeStrainEnergyNeoHookeanNearlyIncompressible](https://github.com/richmondodufisan/purple/blob/main/src/kernels/ComputeStrainEnergyNeoHookeanNearlyIncompressible.C) I:

-compute W
-use finite difference to compute dWdC
-compute the PK2 stress as 2 * dWdC - p*J*C_inverse
-use finite difference calculate the Jacobian, dSdE
-declare the 2nd order PK2 stress and the 4th order Jacobian dS/dE

Again E & C are just C++ RankTwoTensors within this Material,  [ComputeStrainEnergyNeoHookeanNearlyIncompressible](https://github.com/richmondodufisan/purple/blob/main/src/kernels/ComputeStrainEnergyNeoHookeanNearlyIncompressible.C).

The issue I was having is that while the finite difference derivatives worked, the choice of epsilon is problem-dependent. Hence why I wanted to use autodiff (or some other library) where you can define some function, and then calculate the derivative wrt a variable the function depends on.

I am not trying to define the residual or anything like that here right now, I just want to calculate S and dSdE for the new Solid Mechanics module (which is not AD if I understand correctly)

**Reply by GiudGiud** [reply]
Instead of using autodiff to compute W, use our AD numbers.

When you compute W, use AD-numbers. Any term that comes into play in W and depends on C, should be an AD number

You can seed derivatives, even if C is not one of the variables. Like this:
```
ADReal C;
C.derivatives().resize(1);  // this one might take some refinement to get right
C.derivatives()[0] = 1;
auto W = 3 * C;
```
then you should see W.derivatives()[0] to be equal to 3, or dW/dC

**Reply by GiudGiud** [reply]
Though this is pretty uncommon. Usually we only care about the derivatives with regards to the nonlinear variables, not with regards to a tensor C.

Note that my example was with a real C. For a tensor, you will want to see the derivatives of each component of the tensor

**Reply by richmondodufisan** [reply]
I'm still not fully clear on what seeding means, and the structure of the AD variables. Is there a documentation or something I can read through? For example, when you do

```
C.derivatives().resize(1);  // this one might take some refinement to get right
C.derivatives()[0] = 1;
```

What are you resizing? And what is the format of the attribute `.derivatives()`? When you call

`W.derivatives()` 

is it a vector of derivatives of as many seeded variables there are in the function that defines W?

And for a tensor now, would that mean I would need to do something like

```
C.derivatives().resize(3,3)


for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            C.derivatives()[i, j] = 1.0;  // Seed each component's derivative
        }
}
```

And yeah, the reason why I need to be able to calculate derivatives like these is because I will be trying several expressions for W, some of them hard to manually calculate the derivatives of.

**Reply by GiudGiud** [reply]
You can look at the code there for our AD library.
I m afraid we have a lot of examples but not a lot of documentation to teach you how to use it
https://github.com/roystgnr/MetaPhysicL

> What are you resizing? And what is the format of the attribute .derivatives()? When you call

The array of derivatives. It keeps track of the derivatives of the AD number with regards to other values (usually nonlinear variables at a certain dof)

> is it a vector of derivatives of as many seeded variables there are in the function that defines W?

Basically. There is the value of the derivatives, as well as the dof index. Here I am using 0 for the index. You might use more indices for each component of the tensor. 

> And for a tensor now, would that mean I would need to do something like

Probably closer to:
```
C.derivatives().resize(3,3)


for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            C(i,j).derivatives() = (1, i*3 + j);  // Seed each component's derivative
            // 1 because the derivative of Cij wrt to Cij is 1
            // and the index for the index of the component in the derivative vector
        }
}
```

**Reply by richmondodufisan** [reply]
Or would it be something like this?

```
RankTwoTensor ComputeStrainEnergyNeoHookeanNearlyIncompressible::compute_dWdC(const ADReal &mu_0, const ADRankTwoTensor &C)
{
    // Create a seeded tensor
    unsigned int size = 9; // This would be 9 for a 3x3 tensor (???)
    ADRankTwoTensor C_seeded;

    // Loop over each component to seed derivatives
    for (unsigned int i = 0; i < 3; ++i) {
        for (unsigned int j = 0; j < 3; ++j) {
            C_seeded(i, j) = C(i, j);
            C_seeded(i, j).derivatives().resize(size);
            C_seeded(i, j).derivatives()[i * 3 + j] = 1;  // Seed each component's derivative
        }
    }

    // Compute strain energy W using the seeded tensor
    ADReal W = computeStrainEnergy(mu_0, C_seeded);

    // Initialize a new tensor to store derivatives (dW/dC)
    RankTwoTensor dWdC;

    // Extract derivatives with respect to each component of C
    for (unsigned int i = 0; i < 3; ++i) {
        for (unsigned int j = 0; j < 3; ++j) {
            dWdC(i, j) = W.derivatives()[i * 3 + j];  // Get the derivative of W with respect to C(i, j)
        }
    }

    return dWdC;
}
```

**Reply by GiudGiud** [reply]
yes this looks pretty good. 
I think you might need to specify the index as well.If it compiles like this you're good

AD indexing is supposed to be global, which is why the derivatives contain both a value and an index

**Reply by richmondodufisan** [reply]
it compiled but didn't converge unfortunately. I added a cout to see the size of W.derivatives() and it was extremely large, 12816. I don't think the way I set it up is correct.

**Reply by GiudGiud** [reply]
>     ADReal W = computeStrainEnergy(mu_0, C_seeded);

mu_0 does not contain any derivatives with regards to nonlinear variables right?
and neither does any term that comes in computeStrainEnergy right?

We cant get the regular use of AD and our local use mixed up

**Reply by richmondodufisan** [reply]
Nope, mu_0 is just a constant ADReal. And the function is here, could be because of all the matrix operations?

```
ADReal ComputeStrainEnergyNeoHookeanNearlyIncompressible::computeStrainEnergy(const ADReal &mu_0, const ADRankTwoTensor &C)
{
	ADReal J = std::pow(C.det(), 1.0 / 2.0);
	
	// Isochoric Right Cauchy-Green deformation tensor
	ADRankTwoTensor C_iso = std::pow(J, -2.0 / 3.0) * C;
	
	// First invariant, equivalent to lambda_1^2 + lambda_2^2 + lambda_3^2
	ADReal I_1 = C_iso.trace();
	
	ADReal strain_energy = (mu_0/2.0) * (I_1 - 3);
	
	return strain_energy;
}
```

**Reply by GiudGiud** [reply]
this should be fine. However the "det()" operation likely involves every single component, making every component depend on each other

**Reply by richmondodufisan** [reply]
I see. I guess numerical differentiation is the best option here then, I just have to toggle the perturbation value each time. thanks for your help!
